coredb-operator/src/fixtures.rs

```diff
@@ -0,0 +1,150 @@
+//! Helper methods only available for tests
+use crate::{Context, CoreDB, CoreDBSpec, CoreDBStatus, Metrics, COREDB_FINALIZER};
+use assert_json_diff::assert_json_include;
+use futures::pin_mut;
+use http::{Request, Response};
+use hyper::{body::to_bytes, Body};
+use kube::{Client, Resource, ResourceExt};
+use prometheus::Registry;
+use std::sync::Arc;
+use tokio::task::JoinHandle;
+use tower_test::mock::{self, Handle};
+
+impl CoreDB {
+
+    /// A normal test CoreDB
+    pub fn test() -> Self {
+        let mut d = CoreDB::new("testdb", CoreDBSpec::default());
+        d.meta_mut().namespace = Some("testns".into());
+        d.meta_mut().uid = Some("752d59ef-2671-4890-9feb-0097459b18c8".into());
+        d.spec.replicas = 1;
+        d
+    }
+
+    /// Modify a coredb to have the expected finalizer
+    pub fn finalized(mut self) -> Self {
+        self.finalizers_mut().push(COREDB_FINALIZER.to_string());
+        self
+    }
+
+    /// Modify a coredb to have an expected status
+    pub fn with_status(mut self, status: CoreDBStatus) -> Self {
+        self.status = Some(status);
+        self
+    }
+}
+pub struct ApiServerVerifier(Handle<Request<Body>, Response<Body>>);
+
+/// Create a responder + verifier object that deals with the main reconcile scenarios
+///
+/// 1. objects without finalizers will get a finalizer applied (and not call the apply loop)
+/// 2. finalized objects will run through the apply loop
+/// 3. finalized objects "with errors" (i.e. the "illegal" object) will short circuit the apply loop
+/// 4. objects with a deletion timestamp will run the cleanup loop (which will send an event)
+impl ApiServerVerifier {
+    pub fn handle_finalizer_creation(self, coredb_: &CoreDB) -> JoinHandle<()> {
+        let handle = self.0;
+        let coredb = coredb_.clone();
+        tokio::spawn(async move {
+            pin_mut!(handle);
+            let (request, send) = handle.next_request().await.expect("service not called");
+            // We expect a json patch to the specified coredb adding our finalizer
+            assert_eq!(request.method(), http::Method::PATCH);
+            assert_eq!(
+                request.uri().to_string(),
+                format!("/apis/kube.rs/v1/namespaces/testns/coredbs/{}?", coredb.name_any())
+            );
```

Comment:
For example here, the first request that's received is the PATCH to the CoreDB resource, which occurs in the controller here:

https://github.com/CoreDB-io/coredb/blob/7e3237fa9c9698a547f6c7efa7cc29ddb8748f4d/coredb-operator/src/controller.rs#L93-L104